contracts/ronin/fast-finality/FastFinalityTracking.sol

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import { IRoninValidatorSet } from "../../interfaces/validator/IRoninValidatorSet.sol";
+import { IFastFinalityTracking } from "../..//interfaces/IFastFinalityTracking.sol";
+import "../../extensions/collections/HasContracts.sol";
+import "../../utils/CommonErrors.sol";
+
+contract FastFinalityTracking is IFastFinalityTracking, Initializable, HasContracts {
+  /// @dev Mapping from epoch number => consensus address => number of QC vote
+  mapping(uint256 => mapping(address => uint256)) internal _tracker;
+  /// @dev The latest block that tracked the QC vote
+  uint256 internal _latestTrackingBlock;
+
+  modifier oncePerBlock() {
+    if (block.number <= _latestTrackingBlock) {
+      revert ErrOncePerBlock();
+    }
+
+    _latestTrackingBlock = block.number;
+    _;
+  }
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(address _validatorContract) external initializer {
+    _setContract(ContractType.VALIDATOR, _validatorContract);
+  }
+
+  function recordFinality(address[] calldata voters) external override oncePerBlock {
```

Comment:
should check for valid voters address.

Reply:
already checked at nodes

Reply:
what if someone submit tx first that block nodes from calling the function

Reply:
I've just added `onlyCoinbase`